src/merkle/tiered_smt/mod.rs

```diff
@@ -166,6 +166,32 @@ impl TieredSmt {
         TieredSmtProof::new(path, entries).expect("Bug detected, TSMT produced invalid proof")
     }
 
+    /// Returns 
+    /// 1. a partial Merkle tree pre-populated with paths leading up to nodes to be inserted,
+    /// one per key in `keys_to_insert`.
+    /// 2. a map `leaf_hash -> (key, value)` which stores the key/value for keys that already contain a value
+    pub fn get_preinsert_partial_merkle_tree<'a>(
+        &'a self,
+        keys_to_insert: impl IntoIterator<Item = &'a RpoDigest>,
+    ) -> Result<(PartialMerkleTree, BTreeMap<RpoDigest, (RpoDigest, Word)>), MerkleError> {
+        let mut pmt = PartialMerkleTree::default();
+        let mut preimage_map = BTreeMap::new();
+
+        for key in keys_to_insert {
+            let (leaf_index, _) = self.nodes.get_leaf_index(key);
+            let path = self.get_path(leaf_index.into())?;
+            let leaf_node = self.get_node(leaf_index.into())?;
+
+            pmt.add_path(leaf_index.value(), leaf_node, path)?;
+
+            if let Some(v) = self.values.get(key) {
+                preimage_map.insert(leaf_node, (*key, *v));
+            }
```

Comment:
We should also handle the case of hitting a leaf at depth 64. In this case, the leaf node could contain many key-value pairs. So, maybe pre-image map values should be vectors.